src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -2989,47 +3097,100 @@ tx_gateway_frontend::get_all_transactions() {
-          if (!res.has_value()) {
-              if (res.error() == tm_stm::op_status::not_leader) {
-                  co_return tx_errc::not_coordinator;
-              }
-              co_return tx_errc::unknown_server_error;
-          }
+          return stm->read_lock().then([stm](ss::basic_rwlock<>::holder unit) {
+              return stm->get_all_transactions()
+                .then(
+                  [](tm_stm::get_txs_result res)
+                    -> ss::future<return_all_txs_res> {
+                      if (!res.has_value()) {
+                          if (res.error() == tm_stm::op_status::not_leader) {
+                              return ss::make_ready_future<return_all_txs_res>(
+                                return_all_txs_res{tx_errc::not_coordinator});
+                          }
+                          return ss::make_ready_future<return_all_txs_res>(
+                            return_all_txs_res{tx_errc::unknown_server_error});
+                      }
+                      return ss::make_ready_future<return_all_txs_res>(
+                        res.value());
+                  })
+                .finally([u = std::move(unit)] {});
+          });
+      });
+}
 
-          co_return res.value();
+ss::future<tx_gateway_frontend::return_all_txs_res>
+tx_gateway_frontend::get_all_transactions() {
+    auto ntp_meta = _metadata_cache.local().get_topic_metadata(
+      model::tx_manager_nt);
+    if (!ntp_meta) {
+        vlog(
+          txlog.error,
+          "Topic {} doesn't exist in metadata cache",
+          model::tx_manager_nt);
+        co_return tx_errc::partition_not_exists;
+    }
+    tx_gateway_frontend::return_all_txs_res res{{}};
+    co_await ss::do_for_each(
+      ntp_meta->get_assignments(), [this, &res](partition_assignment pa) {
+          auto tx_manager_ntp = model::ntp(
+            model::tx_manager_nt.ns, model::tx_manager_nt.tp, pa.id);
+          return get_all_transactions_for_one_tx_partition(tx_manager_ntp)
+            .then([&res](tx_gateway_frontend::return_all_txs_res ntp_res) {
+                if (res.has_error()) {
+                    return ss::now();
+                }
+                if (ntp_res.has_error()) {
+                    res = ntp_res;
+                    return ss::now();
+                }
+                res.value().insert(
+                  res.value().end(),
+                  ntp_res.value().begin(),
+                  ntp_res.value().end());
+                return ss::now();
+            });
       });
+    co_return res;
 }
 
 ss::future<result<tm_transaction, tx_errc>>
 tx_gateway_frontend::describe_tx(kafka::transactional_id tid) {
-    auto shard = _shard_table.local().shard_for(model::legacy_tm_ntp);
+    auto tx_ntp_opt = _tx_coordinator_ntp_mapper.local().ntp_for(
+      kafka::group_id(tid));
+    if (!tx_ntp_opt) {
+        vlog(
+          txlog.error,
+          "Topic {} doesn't exist in metadata cache,",
```

Comment:
does this message needed update (same as before: not using metadata cache and not searching for topic, but rather transaction).